cEP-0021.md

```diff
@@ -0,0 +1,157 @@
+`pstats` or `snakeviz`.
+
+- `--profile-bears` or `profile_bears` (using `.coafile`) is the main
+  argument to enable profiling. It accepts an additional parameter
+  `Dirname` through which bear writers can specify where to store the dump.
+
+The profiler will start by profiling the `run()` method of bears
+because it consumes most of the bears time. So, this is part where bear writer
+will spend time, as rest of the part like loading the files, collecting the
+settings, etc. are done by coala itself.
+
+## Implementation Details
+
+When user invokes the profiler with either *CLI* or by `coafile`
+then the `execute` method of Bear base class will call the
+`run_bear_from_section` method.
+
+```python
+def run_bear_from_section(self, args, kwargs):
+  ...
+  profile_bears = kwargs.get('profile_bears')
+  if profile_bears is None:
+      self.run(*args, **kwargs)
+  elif:
+      self.profile(run, *args, **kwargs)
+  ...
+```
+
+`run_bear_from_section` method checks whether bear writer invoke the profiler
+interface or not i.e., keyword arguments passed to the Bear contain the
+`profile_bears` key or not.
+
+- If *not* then `run_bear_from_section()` method will directly call
+  the `run()` method of Bear so that coala can continue its normal
+  execution.
+- Else `run_bear_from_section()` method will call `profile()` method which
+  will check the whether the value of `profile_bears` key value is
+  `True` or `Dirname`.
+
+  - If `True` then `profile()` will create a profiler file for every
+    bear of every section with a name `{section_name}{bear_name}.prof` and
+    dump it in a *current working directory.*
+
+    ```python
+    path =  os.getcwd()
+    if str(profile_bears)=='True':
+        filename = '{}_{}.prof'.format(
+            kwargs['section_name'], kwargs['bear_name'])
+        open(filename, 'w')
+        kwargs.pop('section_name')
+        kwargs.pop('bear_name')
+        cProfile.runctx('run(*args, **kwargs)',
+                        globals(), locals(), filename)
+    ```
+
+  - If `Dirname` then `profile()` method will create a profiler file for
+    every bear of every section with a name `{section_name}{bear_name}.prof` and
+    after that it check whether the specified directory exits or not.
+    If exits then it will directly dump the profiler files on the specified
+    directory. If not then it will create a subdirectory name `Dirname`
+    and dump the profiler files on that.
+
+    ```python
+    subdir = str(profile_dump)
+    filepath = os.path.join(path, subdir)
+    if os.path.isdir(filepath):
+        os.chdir(subdir)
+    else:
+        os.mkdir(os.path.join(path, subdir))
+        os.chdir(subdir)
+    filename = '{}_{}.prof'.format(
+        kwargs['section_name'], kwargs['bear_name'])
+    open(filename, 'w')
+    kwargs.pop('section_name')
+    kwargs.pop('bear_name')
+    cProfile.runctx('run(*args, **kwargs)',
+                    globals(), locals(), filename)
+    os.chdir("..")
```

Comment:
This is a profiler, suppose I created a Bear and run the profiler to get the performance and because of newbie bear writer he made some mistake and because of that mistake, it throws some error, so how am I going to predict the what is the exception.?
What can I do for now is handle the exception for now and if any exception is occurring then throw a warning that this exception has occurred.

Reply:
Good point about exception handling, haven't thought about that. Not sure right now what's the best way to go with those^^ Let's wait for the initial implementation try it out, and then decide what we would like to see instead or if we want to keep it.

>  it throws some error, so how am I going to predict the what is the exception.?

Why do you want to predict the exception? Don't predict it, just let it happen :O

Reply:
It was my bad english, I can't even express my feelings with `English` :disappointed: